### ISSUE TITLE
Variables not found in tf 2.5 when calling fit()

### ISSUE THREAD

**Issue by AhmadObeid** [tensorflow/tensorflow#49723]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Google Colab run on Windows 10
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**: N/A
-   **TensorFlow installed from (source or binary)**:  \*
-   **TensorFlow version (use command below)**:  2.5
-   **Python version**: 3.7.10
-   **Bazel version (if compiling from source)**: \*
-   **GCC/Compiler version (if compiling from source)**: \*
-   **CUDA/cuDNN version**: \*
-   **GPU model and memory**: \*
-   **Exact command to reproduce**: see below

\*: For all such information, I am using Google Colab as is, without any modification.

### Description of the problem
I believe this is a problem with tf 2.5, as I was running all codes with no issue for the past few months with tf 2.4.1. The problem happened all of sudden and I came to notice a sudden change in the tf version to 2.5. As the later is the only change, I beleive it must be the reason.

I am building a training model using keras api, everything is straightforward, with Adam chosen as an optimizer. The error happens once I call mdl.fit()

### Source code:
For reference, this is a VAE model. To reproduce, create a colab file and paste the following code as is.
```
seq_shape = (34,4,1)
batch_size = 512
latent_dim = 2

input_seq = keras.Input(shape=seq_shape)
x = layers.Conv2D(80, (5,1),activation='relu')(input_seq)
shape_before_flattening = K.int_shape(x)
x = layers.Flatten()(x)

x = layers.Dense(80,activation='relu')(x)
x = layers.Dense(40,activation='relu')(x)
dense3 = layers.Dense(latent_dim)
z_mean = dense3(x)
z_log_var = layers.Dense(latent_dim)(x)
z = layers.Lambda(sampling)([z_mean, z_log_var])

decoder_input = layers.Input(K.int_shape(z)[1:])
x = layers.Dense(40,activation='relu')(decoder_input)
x = layers.Dense(80,activation='relu')(x)
x = layers.Dense(np.prod(shape_before_flattening[1:]),activation='relu')(x)
x = layers.Reshape(shape_before_flattening[1:])(x)
x = layers.Conv2DTranspose(1, (5,1),activation='sigmoid')(x)
decoder = Model(decoder_input, x)

z_decoded = decoder(z)  
vae = Model(input_seq, z_decoded)


vae.compile(optimizer='adam',loss=[vae_loss])
x = np.ones((100,34,4,1))
vae.fit(x=x,y=x,batch_size=batch_size,shuffle=True,epochs=100)

```
To run the above, the following two functions are needed:
```
def vae_loss(x, z_decoded):
    x = K.flatten(x)
    z_decoded = K.flatten(z_decoded)
    xent_loss =  keras.metrics.binary_crossentropy(x, z_decoded)
    kl_loss =  1e-3 * -0.5 * K.mean( 
    1 + z_log_var - K.square(z_mean) - K.exp(z_log_var), axis=-1)
    return K.mean(xent_loss + kl_loss)

def sampling(args):
    z_mean, z_log_var = args
    epsilon = K.random_normal(shape=(K.shape(z_mean)[0], latent_dim),
    mean=0., stddev=1.)
    return z_mean + K.exp(z_log_var) * epsilon

```
These are all the needed imports:
```
import tensorflow.compat.v1.keras.backend as K
import tensorflow as tf
tf.compat.v1.disable_eager_execution()
import keras
from keras import layers
from keras import backend as K
from keras.models import Model
import numpy as np
```
### Logs
This is the error message I get:

> FailedPreconditionError: 2 root error(s) found.
>   (0) Failed precondition: Could not find variable training_4/Adam/learning_rate. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status=Not found: Resource localhost/training_4/Adam/learning_rate/N10tensorflow3VarE does not exist.
> 	 [[{{node training_4/Adam/Identity/ReadVariableOp}}]]
>   (1) Failed precondition: Could not find variable training_4/Adam/learning_rate. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status=Not found: Resource localhost/training_4/Adam/learning_rate/N10tensorflow3VarE does not exist.
> 	 [[{{node training_4/Adam/Identity/ReadVariableOp}}]]
> 	 [[_arg_input_15_0_0/_359]]
> 0 successful operations.
> 0 derived errors ignored.

Lastly, I noticed that the reported error changes every time I call vae.fit(). The shown log reports not finding leraning_rate, in other instances it was beta_1, beta_2, or the iteration number. 
### Things I already tried:

- Using 'rmsprop' instead of 'adam': did not work. Gives the same error, with differences in the details.
- `from tensorflow.keras.optimizers import Adam`: The error persists

Thanks,

**Comment by UsharaniPagadala** [comment]
@AhmadObeid 

Could you please share the colab gist to reproduce the issue reported here.Thanks

**Comment by AhmadObeid** [comment]
@UsharaniPagadala,
here is the link:
https://colab.research.google.com/drive/1APeQKLFdSczi6FcXnwnOpObocHN4CIM_?usp=sharing
Thank you

**Comment by UsharaniPagadala** [comment]
@AhmadObeid 
I was able to reproduce the code shared with minimal changes in the imports.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/018643b5b05860307d3e764d495e5724/vaegithub_issue.ipynb) here and let us know if it helps.Thanks

**Comment by AhmadObeid** [comment]
@benyeoh  and @UsharaniPagadala,
I confirm that changing some imports as suggested in both of your answers fixes the problem, even with tensorflow 2.5.
Thank you very much for your help.
